libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/InMemoryDataStore.java

```diff
@@ -0,0 +1,205 @@
+import uk.gov.di.ipv.core.library.exceptions.ItemAlreadyExistsException;
+import uk.gov.di.ipv.core.library.persistence.item.PersistenceItem;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class InMemoryDataStore<T extends PersistenceItem> implements DataStore<T> {
+    private static final String KEY_SEPARATOR = "/";
+    private static final ConcurrentMap<String, Map<String, ?>> TABLES = new ConcurrentHashMap<>();
+
+    private final ConcurrentMap<String, T> records;
+    private final Class<T> klass;
+    private final Method partitionKeyMethod;
+    private final Method sortKeyMethod;
+    private final boolean hasSortKey;
+
+    @SuppressWarnings("unchecked") // Rely on callers to use consistent types per table
+    public InMemoryDataStore(String tableName, Class<T> klass) {
+        Method partitionKey = null;
+        Method sortKey = null;
+        for (var method : klass.getMethods()) {
+            if (method.isAnnotationPresent(DynamoDbPartitionKey.class)) {
+                partitionKey = method;
+            }
+            if (method.isAnnotationPresent(DynamoDbSortKey.class)) {
+                sortKey = method;
+            }
+        }
+        if (partitionKey == null) {
+            throw new IllegalArgumentException(
+                    "Missing partition key from class " + klass.getName());
+        }
+
+        this.records =
+                (ConcurrentMap<String, T>)
+                        TABLES.computeIfAbsent(tableName, k -> new ConcurrentHashMap<>());
+        this.klass = klass;
+        this.partitionKeyMethod = partitionKey;
+        this.sortKeyMethod = sortKey;
+        this.hasSortKey = sortKey != null;
+    }
+
+    @Override
+    public void create(T item, ConfigurationVariable tableTtl) {
+        create(item);
+    }
+
+    @Override
+    public void create(T item) {
+        records.put(getKey(item), item);
+    }
+
+    @Override
+    public void createIfNotExists(T item) throws ItemAlreadyExistsException {
+        var key = getKey(item);
+        if (records.putIfAbsent(key, item) != null) {
+            throw new ItemAlreadyExistsException();
+        }
+    }
+
+    @Override
+    public T getItem(String partitionValue, String sortValue) {
+        return records.get(getKey(partitionValue, sortValue));
+    }
+
+    @Override
+    public T getItem(String partitionValue) {
+        return records.get(getKey(partitionValue, null));
+    }
+
+    @Override
+    public T getItem(String partitionValue, boolean warnOnNull) {
+        return getItem(partitionValue);
+    }
+
+    @Override
+    public T getItemByIndex(String indexName, String value) {
+        Method indexMethod = null;
+        for (var method : klass.getMethods()) {
```

Comment:
Part of me is wondering if it's worth checking for secondary index's in the constructor, and creating an index map for them if they do. Then keeping them up to date when writing. But we're probaby never going to have "loads" of items when running locally so it might be overkill.

Reply:
Yeah... I think it's probably not worth it, unless we need this to be super performant. Reflection is 'slow', but not _that_ slow (it'll be orders of magnitude faster than an actual call to dynamodb!)